Add data-driven check of status to Rent History flow

sraby opened the pull request:
This PR introduces a new step in our Rent History Online Form flow that notifies the user if there is any record of a rent stabilized unit existing in their building. It also clarifies language throughout the flow to specify that a rent history will only be mailed to them if their apartment was ever rent stabilized.

sraby commented on frontend/lib/rh/rental-history.tsx:
```diff
@@ -349,9 +436,14 @@ export const getRentalHistoryRoutesProps = (): ProgressRoutesProps => ({
   stepsToFillOut: [
     {
       path: JustfixRoutes.locale.rh.form,
-      exact: true,
```
So odd that this never tripped us up before!

sraby commented on frontend/lib/rh/rental-history.tsx:
```diff
@@ -198,74 +199,139 @@ function RentalHistoryForm(): JSX.Element {
+          <p>
+            <Trans>
+              Your building had{" "}
+              <Plural
+                value={rsInfo.latestUnitCount}
+                one="1 rent stabilized unit"
+                other="# rent stabilized units"
+              />{" "}
```
@toolness for some reason this plural tag isn't getting recognized by lingui when running the extract command, and I can't for the life of me figure out why. Any help would be life changing!

toolness replied:
I think this is because for non-trivial internationalization (e.g. not just simple string substitution), you actually need to run _both_ `yarn lingui:extract` _and_ `yarn lingui:compile`.  This is because at an implementation level, non-trivial i18n is actually accomplished via a JS function that `yarn lingui:compile` generates and puts into the locale bundle file.  Without that pre-compiled functionality, there can be no pluralization or other fancy non-trivial i18n.

toolness replied:
That said, ideally our build pipeline should be running `yarn lingui:compile` whenever it notices that the `.po` files have changed!  That way developers wouldn't have to remember to do this step in such non-trivial situations... I've filed #1780 to address this later.

sraby replied:
@toolness so I've tried compiling and I think the problem is deeper than that? See below— for some reason the plural tag isn't getting written into our messages.po files at all. lmk if I'm missing something

sraby commented on locales/en/messages.po:
```diff
@@ -2087,39 +2100,43 @@ msgstr "Your NoRent letter and important next steps"
 msgstr "Your apartment may be rent stabilized."
 
-#: frontend/lib/data-driven-onboarding/data-driven-onboarding.tsx:317
+#: frontend/lib/rh/rental-history.tsx:138
+msgid "Your building had in {0}, according to property tax documents."
+msgstr "Your building had in {0}, according to property tax documents."
+
```
@toolness see here! This should read "Your building had {0, plural, one...." but instead the whole plural tag is left out!

toolness replied:
holy cow this is truly messed up

sraby replied:
I knoww! Let's wait until next week to figure it out?

toolness replied:
um i found out the problem... or rather i found a solution, i have no idea what the underlying problem is.  Apparently  it has to do with the ORDER in which we import things from `@lingui/macro`... if we do this:

```js
import { Plural, Trans } from "@lingui/macro";
```

the `<Plural>` tag does not work. But if we change it to this:

```js
import { Trans, Plural } from "@lingui/macro";
```

everything works fine?!!!1111

toolness replied:
Oy ok this is logged in https://github.com/lingui/js-lingui/issues/372#issuecomment-637783760 and is apparently a known issue with Lingui.  It's fixed in Lingui v3, which ... whoa appears to have finally been released!!!  I've filed #1782 so we upgrade to it and won't be bitten by this again in the future.

sraby replied:
oh wow amazing, you are really saving the day today!

toolness commented on frontend/lib/rh/rental-history.tsx:
```diff
@@ -306,16 +372,37 @@ function RentalHistoryConfirmation(): JSX.Element {
-          </EnglishOutboundLink>
+          </EnglishOutboundLink>{" "}
+          or by checking out our
+          <LocalizedOutboundLink
+            hrefs={{
+              en:
+                "https://www.justfix.nyc/en/learn/my-landlord-is-overcharging-me",
+              es:
+                "https://www.justfix.nyc/es/learn/my-landlord-is-overcharging-me",
+            }}
+          >
```
Once we have more learning center links, it might be helpful for us to make a `<LearningCenterLink>` component that makes this kind of thing less verbose (it should be more DRY since we know how a learning center URL changes depending on locale).  No need to right now though!

toolness commented on rh/schema.py:
```diff
@@ -37,6 +76,20 @@ class RhForm(DjangoSessionFormMutation):
+
+        full_address = form_data["address"] + ", " + form_data["borough"]
+        bbl, _, _ = lookup_bbl_and_bin_and_full_address(full_address)
+        if bbl:
+            rent_stab_info = process_rent_stab_data(run_rent_stab_sql_query(bbl))
+            request.session[RENT_STAB_INFO_SESSION_KEY] = rent_stab_info
+        return result
```
Hmm, shouldn't there be an `else` clause here that sets the rent stab info to an empty dictionary?  Otherwise if the user already has submitted the form with geosearch working, but then goes back and changes their address, but this time geosearch _doesn't_ work, the stale rent stab info for their previously-entered address will still be in the session.

**Update:** rather than an `else` clause, it might actually be easier to just set `rent_stab_info` to `None` outside of the clause and set the session key to `rent_stab_info` after the logic is over, since there are a number of ways that `rent_stab_info` could end up being `None` (either because geosearch failed or because NYCDB isn't configured).

We should also add tests for this mutation to make sure the rent stab info is properly updated in either case!

toolness commented on rh/schema.py:
```diff
@@ -76,6 +129,29 @@ def perform_mutate(cls, form, info):
+        request = info.context
+        kwargs = request.session.get(RENT_STAB_INFO_SESSION_KEY, {})
+        if kwargs:
+            return RhRentStabData(**kwargs)
```
This case should have test coverage... we could unify it into the test for the mutation, that is how many of my GraphQL tests work.

sraby replied:
Slated some time on Monday for us to work through this together!

toolness commented on rh/schema.py:
```diff
@@ -26,6 +37,34 @@ def get_slack_notify_text(rhr: models.RentalHistoryRequest) -> str:
+
+def process_rent_stab_data(raw_data: Optional[Dict[str, Any]]) -> Optional[Dict[str, Any]]:
+    if not raw_data:
+        return BLANK_RENT_STAB_INFO
```
Er, shouldn't this return `None` if `raw_data` is `None`?  It seems like `raw_data` is only going to be `None` if NYCDB isn't configured... hmm.

Actually it feels confusing to have a function that takes an optional argument for the sole purpose of returning `None` when it's passed `None`.  Maybe this function should take a non-optional `Dict[str, Any]` and always return a non-optional `Dict[str, Any]` instead, and the caller should be in charge of only calling the function if it has actual `raw_data` to provide?

It seems like perhaps a _third_ helper function would be useful here, one called `get_rent_stab_data_for_bbl` which takes a BBL and returns an `Optional[Dict[str, Any]]`...

toolness commented on rh/schema.py:
```diff
@@ -26,6 +37,34 @@ def get_slack_notify_text(rhr: models.RentalHistoryRequest) -> str:
 
 
+def run_rent_stab_sql_query(bbl: str) -> Optional[Dict[str, Any]]:
+    sql_query = """
+        select uc2007, uc2008, uc2009, uc2010, uc2011, uc2012, uc2013,
+             uc2014, uc2015, uc2016, uc2017, uc2018, uc2019
+        from rentstab
+        full join rentstab_v2 using(ucbbl)
+        where ucbbl = %(bbl)s
+    """
+    if not settings.NYCDB_DATABASE:
+        return None
+
+    with connections[settings.NYCDB_DATABASE].cursor() as cursor:
+        cursor.execute(sql_query, {"bbl": bbl})
+        json_result = list(generate_json_rows(cursor))
+        if not json_result:
+            return BLANK_RENT_STAB_INFO
+        return json_result[0]
```
Ideally I think it'd be good if, in the future, we could leverage the `nycdb` python package to scaffold some of this stuff for us, so that we could actually test this code.  That's beyond the scope of this PR but it'd be good to file an issue reminding us to do it later...

toolness reviewed (changes requested):
This is a good start!  I've left a few comments about some things that should be addressed, let me know if you want to go over these in a video chat or anything.

sraby commented on frontend/lib/rh/rental-history.tsx:
```diff
@@ -306,16 +372,37 @@ function RentalHistoryConfirmation(): JSX.Element {
-          Rent History is an important document—it shows the registered rents in
-          your apartment since 1984. You can learn more about it and how it can
-          help you figure out if you’re being overcharged on rent at the{" "}
+          If your apartment is currently rent stabilized— or has been at any
+          point in the past— you should receive your Rent History in the mail in
```
oooh, grammar thing?

sraby commented:
Ok @toolness, I addressed all of your comments above (with the exception of writing a "unified" test for the graphql mutation and `resolve_rent_stab_info` function. 

Most notably, I refactored our logic in schema.py so that we are ALWAYS resetting our session data on every mutation, and we are appropriately assigning a None value, a blank object, or a populated object as the value of the rent_stab_info field. See below:

sraby commented on rh/schema.py:
```diff
@@ -26,6 +37,46 @@ def get_slack_notify_text(rhr: models.RentalHistoryRequest) -> str:
+    return BLANK_RENT_STAB_INFO
+
+
+def get_rent_stab_info_for_bbl(bbl: str) -> Optional[Dict[str, Any]]:
+    # Case 1: No connection to the database
+    if not settings.NYCDB_DATABASE:
+        return None
+
+    raw_data = run_rent_stab_sql_query(bbl)
+
+    # Case 2: We connected to the database, but no RS data was found
+    if not raw_data:
+        return BLANK_RENT_STAB_INFO
+
+    # Case 3: We connected to the database, and RS data was found
+    else:
+        return process_rent_stab_data(raw_data)
+
```
New helper function here spells out our "three cases" of getting rent stab info from a bbl.

toolness replied:
Nice!  Based on its usage, though, I think there's a fourth case--the case where `bbl` is an empty string because the geosearch service wasn't available?  So maybe change line 66 to `if not (bbl and settings.NYCDB_DATABASE):`?

sraby replied:
good point!

sraby commented on rh/schema.py:
```diff
@@ -37,6 +88,18 @@ class RhForm(DjangoSessionFormMutation):
+
+        full_address = form_data["address"] + ", " + form_data["borough"]
+        bbl, _, _ = lookup_bbl_and_bin_and_full_address(full_address)
+        request.session[RENT_STAB_INFO_SESSION_KEY] = get_rent_stab_info_for_bbl(bbl)
```
now, we reset rent stab info on every mutate, addressing your comment earlier about having stale data present

toolness commented on rh/schema.py:
```diff
@@ -26,6 +37,48 @@ def get_slack_notify_text(rhr: models.RentalHistoryRequest) -> str:
+
+    raw_data = run_rent_stab_sql_query(bbl)
+
+    # Case 2: We connected to the database, but no RS data was found
+    if not raw_data:
+        return BLANK_RENT_STAB_INFO
```
Just noticed that this doesn't have test coverage--would be nice to add a test for it, or file an issue to add such a test lf we want to get i18n done asap!

sraby replied:
Just added this test case in!

toolness reviewed (approved):
This looks good!  Added just one comment.
